src/lib/dovetail_audio.ts

```diff
@@ -0,0 +1,221 @@
+import {AdzerkRequest, AdzerkResponse} from './adzerk';
+import {AdzerkFetcher} from './adzerk_fetcher';
+import {ExtendableAudio} from '../javascript/extendable_audio';
+import {DovetailFetcher, DovetailResponse, ArrangementEntry} from './dovetail_fetcher';
+
+type AllUnion = [
+  DovetailResponse | PromiseLike<DovetailResponse>,
+  AdzerkResponse | PromiseLike<AdzerkResponse>
+];
+
+type AllResult = [DovetailResponse, AdzerkResponse];
+
+interface Arrangement {
+  entries: ArrangementEntry[];
+  duration?: number;
+}
+
+function sumDuration(collector: number, entry: {duration?: number}) {
+  return collector + entry.duration || 0;
+}
+
+const DURATION_CHANGE = 'durationchange';
+const TIME_UPDATE = 'timeupdate';
+const ENDED = 'ended';
+const AD_START = 'adstart';
+const AD_END = 'adend';
+
+class DovetailEvent {
+  static build(eventName: string, audio: DovetailAudio, extras?: {}) {
+    let event = new Event(eventName);
+    Object.defineProperty(event, 'target', {
+      value: audio,
+      writable: false
+    });
+    Object.defineProperty(event, 'currentTarget', {
+      value: audio,
+      writable: false
+    });
+    if (extras) {
+      for (let property in extras) {
+        if (extras.hasOwnProperty(property)) {
+          event[property] = extras[property];
+        }
+      }
+    }
+    return event;
+  }
+}
```

Comment:
We basically do this to mask that the events are hosted by a DocumentFragment so that the target appears to be the actual DovetailAudio element (so in your event handler you can do `if (e.target == audio)` and have it work correctly.)